File: generated_codes/chat_agents/chat-agents-cpp/bst/gemini/output2_bst.cpp
#pragma once
#include "../utils/SetADT.hpp"
#include <atomic>
#include <climits>
#include <cstdint>
#include <vector>

class ConcurrentDataStructure : public SetADT {
private:
    struct Node {
        int val;
        std::atomic<Node*> left;
        std::atomic<Node*> right;

        explicit Node(int v)
            : val(v), left(nullptr), right(nullptr) {}
    };

    Node* root;

    static bool is_marked_ref(Node* p) {
        return (reinterpret_cast<uintptr_t>(p) & 1ULL) != 0;
    }

    static Node* get_unmarked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) & ~1ULL);
    }

    static Node* get_marked_ref(Node* p) {
        return reinterpret_cast<Node*>(reinterpret_cast<uintptr_t>(p) | 1ULL);
    }

    struct SearchResult {
        Node* parent;
        Node* node;
        std::atomic<Node*>* parent_link;
    };

    bool is_logically_deleted(Node* node) const {
        Node* l = node->left.load(std::memory_order_acquire);
        Node* r = node->right.load(std::memory_order_acquire);
        return is_marked_ref(l) && is_marked_ref(r);
    }

    bool mark_node(Node* node) {
        while (true) {
            Node* left = node->left.load(std::memory_order_acquire);
            if (!is_marked_ref(left)) {
                if (!node->left.compare_exchange_weak(
                        left,
                        get_marked_ref(left),
                        std::memory_order_acq_rel,
                        std::memory_order_acquire)) {
                    continue;
                }
            }
            break;
        }

        while (true) {
            Node* right = node->right.load(std::memory_order_acquire);
            if (!is_marked_ref(right)) {
                if (!node->right.compare_exchange_weak(
                        right,
                        get_marked_ref(right),
                        std::memory_order_acq_rel,
                        std::memory_order_acquire)) {
                    continue;
                }
            }
            break;
        }

        return true;
    }

    bool help_unlink(Node* parent, std::atomic<Node*>* parent_link, Node* node) {
        if (!parent_link) {
            return false;
        }

        Node* left = get_unmarked_ref(node->left.load(std::memory_order_acquire));
        Node* right = get_unmarked_ref(node->right.load(std::memory_order_acquire));

        if (left && right) {
            return false;
        }

        Node* replacement = left ? left : right;

        Node* expected = node;
        return parent_link->compare_exchange_strong(
            expected,
            replacement,
            std::memory_order_acq_rel,
            std::memory_order_acquire);
    }

    SearchResult search(int key) {
        retry:
        Node* parent = nullptr;
        std::atomic<Node*>* parent_link = nullptr;

        Node* curr = root;

        while (curr) {
            if (is_logically_deleted(curr)) {
                if (parent && parent_link) {
                    help_unlink(parent, parent_link, curr);
                    goto retry;
                }
            }

            if (curr->val == key) {
                return {parent, curr, parent_link};
            }

            parent = curr;

            if (key < curr->val) {
                parent_link = &curr->left;
                curr = get_unmarked_ref(
                    curr->left.load(std::memory_order_acquire));
            } else {
                parent_link = &curr->right;
                curr = get_unmarked_ref(
                    curr->right.load(std::memory_order_acquire));
            }
        }

        return {parent, nullptr, parent_link};
    }

    void destroy(Node* node) {
        if (!node) {
            return;
        }

        std::vector<Node*> stack;
        stack.push_back(node);

        while (!stack.empty()) {
            Node* curr = stack.back();
            stack.pop_back();

            Node* left =
                get_unmarked_ref(curr->left.load(std::memory_order_relaxed));
            Node* right =
                get_unmarked_ref(curr->right.load(std::memory_order_relaxed));

            if (left) {
                stack.push_back(left);
            }

            if (right) {
                stack.push_back(right);
            }

            delete curr;
        }
    }

public:
    ConcurrentDataStructure() {
        Node* min_node = new Node(INT_MIN);
        Node* max_node = new Node(INT_MAX);
        min_node->right.store(max_node, std::memory_order_release);
        root = min_node;
    }

    ~ConcurrentDataStructure() override {
        destroy(root);
    }

    bool contains(int key) override {
        Node* curr = root;

        while (curr) {
            if (!is_logically_deleted(curr) && curr->val == key) {
                return true;
            }

            if (key < curr->val) {
                curr = get_unmarked_ref(
                    curr->left.load(std::memory_order_acquire));
            } else {
                curr = get_unmarked_ref(
                    curr->right.load(std::memory_order_acquire));
            }
        }

        return false;
    }

    bool add(int key) override {
        if (key == INT_MIN || key == INT_MAX) {
            return false;
        }

        while (true) {
            SearchResult s = search(key);

            if (s.node && !is_logically_deleted(s.node)) {
                return false;
            }

            Node* parent = s.parent;
            std::atomic<Node*>* link = s.parent_link;

            if (!parent || !link) {
                continue;
            }

            Node* new_node = new Node(key);

            Node* expected = nullptr;

            if (link->compare_exchange_strong(
                    expected,
                    new_node,
                    std::memory_order_acq_rel,
                    std::memory_order_acquire)) {
                return true;
            }

            delete new_node;
        }
    }

    bool remove(int key) override {
        if (key == INT_MIN || key == INT_MAX) {
            return false;
        }

        while (true) {
            SearchResult s = search(key);

            if (!s.node) {
                return false;
            }

            Node* node = s.node;

            Node* left =
                get_unmarked_ref(node->left.load(std::memory_order_acquire));
            Node* right =
                get_unmarked_ref(node->right.load(std::memory_order_acquire));

            if (left && right) {
                Node* succ_parent = node;
                std::atomic<Node*>* succ_link = &node->right;
                Node* succ = right;

                while (true) {
                    Node* next =
                        get_unmarked_ref(succ->left.load(std::memory_order_acquire));

                    if (!next) {
                        break;
                    }

                    succ_parent = succ;
                    succ_link = &succ->left;
                    succ = next;
                }

                int succ_val = succ->val;

                if (!remove(succ_val)) {
                    continue;
                }

                node->val = succ_val;
                return true;
            }

            mark_node(node);

            if (help_unlink(s.parent, s.parent_link, node)) {
                return true;
            }

            if (!s.parent) {
                return false;
            }
        }
    }
};

File: generated_codes/chat_agents/chat-agents-java/bst/chatgpt_java/output2_bst.java
package com.example.Sets;

import com.example.utils.SetADT;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class ConcurrentDataStructure implements SetADT {
private static final class Node {
final int key;
final AtomicMarkableReference<Node> left;
final AtomicMarkableReference<Node> right;


    Node(int key) {
        this.key = key;
        this.left = new AtomicMarkableReference<>(null, false);
        this.right = new AtomicMarkableReference<>(null, false);
    }

    boolean isMarked() {
        return right.isMarked();
    }
}

private final AtomicMarkableReference<Node> root;

public ConcurrentDataStructure() {
    this.root = new AtomicMarkableReference<>(null, false);
}

@Override
public boolean add(int key) {
    Node newNode = new Node(key);

    while (true) {
        Node r = root.getReference();
        if (r == null) {
            if (root.compareAndSet(null, newNode, false, false)) {
                return true;
            }
            continue;
        }

        Window w = find(key);
        Node parent = w.parent;
        Node curr = w.curr;

        if (curr != null && curr.key == key && !curr.isMarked()) {
            return false;
        }

        if (parent == null) {
            continue;
        }

        AtomicMarkableReference<Node> childRef = key < parent.key ? parent.left : parent.right;
        Node observed = childRef.getReference();

        if (observed == curr && !childRef.isMarked() && !parent.isMarked()) {
            if (childRef.compareAndSet(curr, newNode, false, false)) {
                return true;
            }
        }
    }
}

@Override
public boolean remove(int key) {
    while (true) {
        Window w = find(key);
        Node parent = w.parent;
        Node curr = w.curr;

        if (curr == null || curr.key != key || curr.isMarked()) {
            return false;
        }

        Node rightChild = curr.right.getReference();
        if (curr.right.compareAndSet(rightChild, rightChild, false, true)) {
            // Node has been marked
            helpUnlink(parent, curr);
            return true;
        }
    }
}

@Override
public boolean contains(int key) {
    Node curr = root.getReference();

    while (curr != null) {
        if (curr.isMarked()) {
            curr = key < curr.key ? curr.left.getReference() : curr.right.getReference();
            continue;
        }

        if (key == curr.key) {
            return true;
        }

        curr = key < curr.key ? curr.left.getReference() : curr.right.getReference();
    }

    return false;
}

private Window find(int key) {
    retry:
    while (true) {
        Node parent = null;
        Node curr = root.getReference();

        while (curr != null) {
            if (curr.isMarked()) {
                if (!helpUnlink(parent, curr)) {
                    continue retry;
                }

                curr = parent == null
                        ? root.getReference()
                        : (key < parent.key ? parent.left.getReference() : parent.right.getReference());
                continue;
            }

            if (key == curr.key) {
                return new Window(parent, curr);
            }

            parent = curr;
            curr = key < curr.key ? curr.left.getReference() : curr.right.getReference();
        }

        return new Window(parent, null);
    }
}

private boolean helpUnlink(Node parent, Node curr) {
    Node leftChild = curr.left.getReference();
    Node rightChild = curr.right.getReference();

    if (leftChild != null && rightChild != null) {
        return true;
    }

    Node replacement = leftChild != null ? leftChild : rightChild;

    if (parent == null) {
        return root.compareAndSet(curr, replacement, false, false) || root.getReference() != curr;
    }

    if (parent.isMarked()) {
        return false;
    }

    AtomicMarkableReference<Node> link = curr.key < parent.key ? parent.left : parent.right;
    return link.compareAndSet(curr, replacement, false, false) || link.getReference() != curr;
}

private static final class Window {
    final Node parent;
    final Node curr;

    Window(Node parent, Node curr) {
        this.parent = parent;
        this.curr = curr;
    }
}


}
